data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/PeerForwarder.java

```diff
@@ -85,69 +93,70 @@ public PeerForwarder(final PluginSetting pluginSetting) {
+            final String dataPrepperIp = hashRing.getServerIp(traceId).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
+            spansByEndPoint.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).addAll(entry.getValue());
         }
 
-        final List<Record<ExportTraceServiceRequest>> recordsToProcessLocally = new ArrayList<>();
-        final List<CompletableFuture<Record>> forwardedRequestFutures = new ArrayList<>();
+        final List<Record<Span>> recordsToProcessLocally = new ArrayList<>();
+        final Map<CompletableFuture<ExportTraceServiceRequest>, List<Span>> forwardedRequestFuturesToSpans = new HashMap<>();
 
-        for (final Map.Entry<String, List<ResourceSpans>> entry : groupedRS.entrySet()) {
+        for (final Map.Entry<String, List<Span>> entry : spansByEndPoint.entrySet()) {
             final TraceServiceGrpc.TraceServiceBlockingStub client = getClient(entry.getKey());
+            if (isLocalClient(client)) {
+                recordsToProcessLocally.addAll(entry.getValue().stream().map(Record::new).collect(Collectors.toList()));
+                continue;
+            }
 
             // Create ExportTraceRequest for storing single batch of spans
             ExportTraceServiceRequest.Builder currRequestBuilder = ExportTraceServiceRequest.newBuilder();
-            int currSpansCount = 0;
-            for (final ResourceSpans rs : entry.getValue()) {
-                final int rsSize = PeerForwarderUtils.getResourceSpansSize(rs);
-                if (currSpansCount >= maxNumSpansPerRequest) {
+            List<Span> currBatchSpans = new ArrayList<>();
+            for (final Span span : entry.getValue()) {
+                final ResourceSpans rs;
+                try {
+                    rs = oTelProtoEncoder.convertToResourceSpans(span);
+                    currRequestBuilder.addResourceSpans(rs);
+                    currBatchSpans.add(span);
+                } catch (UnsupportedEncodingException | DecoderException e) {
+                    LOG.error("failed to encode span with spanId: {} into opentelemetry-protobuf, span will be exported locally.",
+                            span.getSpanId(), e);
+                    recordsToProcessLocally.add(new Record<>(span));
+                }
+                if (currBatchSpans.size() >= maxNumSpansPerRequest) {
                     final ExportTraceServiceRequest currRequest = currRequestBuilder.build();
-                    if (isLocalClient(client)) {
-                        recordsToProcessLocally.add(new Record<>(currRequest));
-                    } else {
-                        forwardedRequestFutures.add(processRequest(client, currRequest));
-                    }
+                    forwardedRequestFuturesToSpans.put(processRequest(client, currRequest), currBatchSpans);
                     currRequestBuilder = ExportTraceServiceRequest.newBuilder();
-                    currSpansCount = 0;
+                    currBatchSpans = new ArrayList<>();
```

Comment:
Thanks for refactoring this entire method. I like the approach you have taken. I think it is much easier to follow along. I think we could push this even further by pull some functionality out into separate methods but we are working on replacing this with the core peer forwarder soon.

Reply:
Will doublecheck. I think this peer-forwarder plugin will still stand for a while.

Reply:
I did not find repeated code to refactor here. Specific suggestions are welcome.